drivers/i2c/meson/i2c.c

```diff
@@ -414,8 +410,52 @@ static inline void i2c_load_tokens(volatile struct i2c_regs *regs)
     // we have already cleared the registers and it has a value of zero.
 
     while (tk_offset < 16 && wdata_offset < 8 && rdata_offset < 8 && request_data_offset < i2c_ifState.curr_request_len) {
-        // Get the SoC specific token
-        uint8_t meson_token = i2c_token_convert(tokens[request_data_offset]);
+        LOG_DRIVER("request_data_offset : 0x%lx, tk_offset: 0x%lx, wdata_offset: 0x%lx, rdata_offset: 0x%lx\n",
+                   request_data_offset, tk_offset, wdata_offset, rdata_offset);
+
+        // Discover next operation
+        uint8_t meson_token, data;
+        if (i2c_ifState.rw_remaining == 0) {
+            LOG_DRIVER("Accepting new token...\n");
+            // Get meson_token if no read/write is in progress.
+            meson_token = i2c_token_convert(tokens[request_data_offset]);
+
+            // Grab buffer length, if appropriate.
+            if (meson_token == MESON_I2C_TOKEN_ADDR_WRITE || meson_token == MESON_I2C_TOKEN_ADDR_READ) {
+                LOG_DRIVER("Beginning RW operation!\n");
+                // R/W buffer incoming: |RD/WR|LEN|DATA0|DATA1| ... |DATA[LEN-1]|
+                uint8_t buff_length = tokens[request_data_offset + 1];
+
+                // Set interface state
+                i2c_ifState.rw_remaining = buff_length;
+                i2c_ifState.data_direction = (meson_token == MESON_I2C_TOKEN_ADDR_WRITE)
+                                             ? DATA_DIRECTION_WRITE : DATA_DIRECTION_READ;
+
+                LOG_DRIVER("DD=%d    REM=%d\n", i2c_ifState.data_direction, i2c_ifState.rw_remaining);
+
+                // IMPORTANT: skip buffer length byte!
+                request_data_offset++;
+            }
+
+        } else {
+            LOG_DRIVER("Resuming in-progress read/write. DD=%d remaining=%d\n",
+                       i2c_ifState.data_direction, i2c_ifState.rw_remaining);
+            // We are in the middle of a read or write. Pick up where we left off
+            if (i2c_ifState.rw_remaining == 1 && i2c_ifState.data_direction == DATA_DIRECTION_READ) {
+                // Write data end on last byte of a read.
+                meson_token = MESON_I2C_TOKEN_DATA_END;
+                // TODO: @mattr potential fencepost error? Seems correct to me though.
```

Comment:
We try to put TODOs in an issue or resolve them before merging, they shouldn't be in the code.